util/fpga/cw_spiflash.py

```diff
@@ -0,0 +1,273 @@
+          fpga: CW310/305 to be programmed, a ``cw.capture.targets.CW310/305`` instance.
+          board: can be ``CW310`` or ``CW305`` to distinguish the different board types.
+        """
+
+        # Change pin mapping to CW310 board.
+        if board == "CW310":
+            self._PIN_SCK = 'USB_SPI_SCK'
+            self._PIN_SDI = 'USB_SPI_COPI'
+            self._PIN_SDO = 'USB_SPI_CIPO'
+            self._PIN_CS = 'USB_SPI_CS'
+            self._PIN_TRST = 'USB_A17'
+            self._PIN_SRST = 'USB_A18'
+            self._PIN_JTAG_SPI = 'USB_A19'
+            self._PIN_BOOTSTRAP = 'USB_A16'
+
+        # Configure JTAG and bootstrap pins.
+        self._fpga_io = fpga.gpio_mode()
+        self._fpga_io.pin_set_output(self._PIN_TRST)
+        self._fpga_io.pin_set_output(self._PIN_SRST)
+        self._fpga_io.pin_set_output(self._PIN_JTAG_SPI)
+        self._fpga_io.pin_set_output(self._PIN_BOOTSTRAP)
+        self._fpga_io.pin_set_state(self._PIN_TRST, 1)
+        self._fpga_io.pin_set_state(self._PIN_SRST, 1)
+        self._fpga_io.pin_set_state(self._PIN_JTAG_SPI, 1)
+        self._fpga_io.pin_set_state(self._PIN_BOOTSTRAP, 0)
+        # Initialize SPI pins.
+        self._fpga_io.spi1_setpins(sck=self._PIN_SCK,
+                                   sdo=self._PIN_SDI,
+                                   sdi=self._PIN_SDO,
+                                   cs=self._PIN_CS)
+        self._fpga_io.spi1_enable(True)
+
+    def _reset_opentitan(self):
+        self._fpga_io.pin_set_state(self._PIN_JTAG_SPI, 1)
+        self._fpga_io.pin_set_state(self._PIN_SRST, 0)
+        time.sleep(self._BOOTSTRAP_DELAY)
+        self._fpga_io.pin_set_state(self._PIN_SRST, 1)
+        time.sleep(self._BOOTSTRAP_DELAY)
+
+    def __enter__(self):
+        """Starts bootstrapping."""
+        self._fpga_io.pin_set_state(self._PIN_BOOTSTRAP, 1)
+        self._reset_opentitan()
+        self._fpga_io.pin_set_state(self._PIN_JTAG_SPI, 0)
+        time.sleep(self._BOOTSTRAP_DELAY)
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Ends bootstrapping."""
+        self._fpga_io.pin_set_state(self._PIN_BOOTSTRAP, 0)
+        self._fpga_io.pin_set_state(self._PIN_JTAG_SPI, 1)
+        time.sleep(self._BOOTSTRAP_DELAY)
+
+    def transfer(self, frame):
+        """Transmits a frame over SPI and receives the acknowledgement for the previous frame."""
+        # Wait longer after the first frame since it triggers a flash erase operation.
+        if frame.is_second_frame:
+            time.sleep(self._SECOND_FRAME_DELAY)
+        else:
+            time.sleep(self._INTER_FRAME_DELAY)
+        print(f'Transferring {frame}.')
+        # Acknowledgement is the same size as digests.
+        return bytes(
+            self._fpga_io.spi1_transfer(
+                bytes(frame))[:_SpiFlashFrame.DIGEST_SIZE])
+
+
+class SPIProgrammer:
+    """Class for programming OpenTitan over the SPI interface of SAM3X/U on CW310/305."""
```

Comment:
This hardly has any state. Maybe run() should be a free function...?

Reply:
When this was initially implemented, the `OpenTitan` class expected the programmer to have a `run` method and this class was written to conform to that interface.